src/Algebra/Primality.agda

```diff
@@ -0,0 +1,55 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Definitions of irreducibility, primality, coprimality.
+------------------------------------------------------------------------
+
+-- You're unlikely to want to use this module directly. Instead you
+-- probably want to be importing the appropriate module from
+-- `Algebra.Properties.CancellativeCommutativeSemiring`.
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Algebra.Core using (Op₂)
+import Algebra.Divisibility
+open import Data.Product using (_×_)
+open import Data.Sum using (_⊎_)
+open import Function using (flip)
+open import Level using (_⊔_)
+open import Relation.Binary using (Rel; Symmetric)
+open import Relation.Nullary using (¬_)
+open import Relation.Unary using (Pred)
+
+module Algebra.Primality
+  {a ℓ} {A : Set a} (_≈_ : Rel A ℓ) (_*_ : Op₂ A) (0# 1# : A)
+  where
+
+open import Algebra.Divisibility _≈_ _*_
+
+------------------------------------------------------------------------------
+-- Definitions
+
+private
+  infix 4 _≉_
+
+  _≉_ : Rel A _
+  x ≉ y = ¬ (x ≈ y)
+
+Irreducible : Pred A (a ⊔ ℓ)
+Irreducible p = p ∤ 1#  ×  (∀ {x y} → (p ≈ (x * y)) → x ∣ 1# ⊎ y ∣ 1#)
```

Comment:
No brackets needed for `p ≈ (x * y)`

Reply:
Following your record requirement, I set
```
record Irreducible (p : A) : Set (a ⊔ ℓ) where
  constructor irredᶜ
  field
    ∤1                     : p ∤ 1#
    chooseInvertibleFactor : ∀ {x y} → p ≈ (x * y) → x ∣ 1# ⊎ y ∣ 1#
```
As to parentheses in the last line, Agda insists of them, because the priorities are not visible.